### PR TITLE
Add back dataset size limitation

### LINKED ISSUE #2777: [TODO] re-implement limit_train/dev/test flags
Add code to ensure the flags below work.
 
f.DEFINE_integer('limit_train', 0, 'maximum number of elements to use from train set - 0 means no limit')
f.DEFINE_integer('limit_dev', 0, 'maximum number of elements to use from validation set- 0 means no limit')
f.DEFINE_integer('limit_test', 0, 'maximum number of elements to use from test set- 0 means no limit')

### PR DESCRIPTION
**Pull request by lissyx**
Fixes #2777

### CODE REVIEW

File: util/feeding.py [hunk 1]
```diff
@@ -109,14 +109,15 @@ def to_sparse_tuple(sequence):
...
+                   limit=-1,
                    enable_cache=False,
                    cache_path=None,
                    train_phase=False,
                    exception_box=None,
                    process_ahead=None,
                    buffering=1 * MEGABYTE):
     def generate_values():
-        samples = samples_from_files(sources, buffering=buffering, labeled=True)
+        samples = samples_from_files(sources, buffering=buffering, labeled=True, sample_limit=limit)
```
Comment (tilmankamp) [inline util/feeding.py]: If we won't go for the above proposal: Why not applying the limit within the following loop (exiting it once the limit is reached)?

Yes, this has the disadvantage of still having to load the complete "lists" of the sample sources (SDB offset table, CSV rows), but on the other hand it would be:
- less intrusive and therefore easier to maintain and 
- far more logical in regards to the outcome: Limiting the combination of all samples instead of limiting each source

File: util/feeding.py [hunk 2]
```diff
@@ -109,14 +109,15 @@ def to_sparse_tuple(sequence):
 
 def create_dataset(sources,
                    batch_size,
+                   limit=-1,
```
Comment (tilmankamp) [inline util/feeding.py]: If I remember correctly, the limit was once introduced in combination with an ability to start from a certain sample offset or to reverse the data-set. This allowed us to test for maximum batch size (testing the large samples only) without having to run a full epoch.

We should really make it `None`-able (no limit) and have negative numbers take samples from the end of the collections.

Comment (lissyx) [reply util/feeding.py]: > f I remember correctly, the limit was once introduced in combination with an ability to start from a certain sample offset or to reverse the data-set.

Show me the code, because as much as I could find:
 - this has been broken for so long
 - this was basically what I implemented



> We should really make it `None`-able (no limit) and have negative numbers take samples from the end of the collections.

I don't get your point.

File: util/feeding.py (hunk 1 above)
Comment (lissyx) [reply util/feeding.py]: > far more logical in regards to the outcome: Limiting the combination of all samples instead of limiting each source

You suggested the syntax limiting per-files, so I'm lost here.



> less intrusive and therefore easier to maintain and

Much much painful to test, though, because `create_dataset` produces tensorflow data ops.

Comment (lissyx) [reply util/feeding.py]: https://github.com/mozilla/DeepSpeech/issues/2777#issuecomment-603392850

File: util/feeding.py (hunk 2 above)
Comment (tilmankamp) [reply util/feeding.py]: At the moment samples are taken from the beginning of the collections (they are typically sorted shortest to largest). So a limit of e.g. 10 samples takes the first 10 samples (typically the shortest ones) of each set. In my proposal -10 would allow the user to take the 10 last samples (the largest ones) of each set instead.

File: util/feeding.py (hunk 1 above)
Comment (reuben) [reply util/feeding.py]: FWIW for testing a dataset, it shouldn't be that hard. In eager mode, just `list(dataset)` will do what you expect. Example: https://gist.github.com/reuben/81ac06018a6bcc581a48d08c51a01a5b

File: util/feeding.py (hunk 2 above)
Comment (reuben) [reply util/feeding.py]: I think he means make it equivalent to Python slicing syntax: `--limit N` means taking `dataset[:N]`. But with negative values this would not be useful for testing maximum batch sizes. For that you'd have to do `dataset[-N:]`.

**Comment by reuben** [comment]
If there's no way to specify per-file limits, then the limit should be global. Having a single `--limit_train` flag that applies separately for each CSV is very counter intuitive.

File: util/feeding.py (hunk 1 above)
Comment (lissyx) [reply util/feeding.py]: > FWIW for testing a dataset, it shouldn't be that hard. In eager mode, just `list(dataset)` will do what you expect. Example: https://gist.github.com/reuben/81ac06018a6bcc581a48d08c51a01a5b

Yes, but that's starting to be quite a lot of setup for unit testing :/

File: util/feeding.py (hunk 2 above)
Comment (lissyx) [reply util/feeding.py]: > I think he means make it equivalent to Python slicing syntax: `--limit N` means taking `dataset[:N]`. But with negative values this would not be useful for testing maximum batch sizes. For that you'd have to do `dataset[-N:]`.

`Interleaved` does not support `reversed()` and `itertools.islice` does not support negative values. Since we `yield` from `SDB` or `CSV` classes within `__iter__` I don't see how we can handle negative values to pull from the end of the dataset without being more intrusive into those classes.

File: util/sample_collections.py [hunk 3]
```diff
@@ -355,11 +355,14 @@ def samples_from_files(filenames, buffering=BUFFER_SIZE, labeled=None):
...
         If None: Reads util.sample_collections.LabeledSample instances from sources with transcripts and
         util.audio.Sample instances from sources with no transcripts.
+    reverse : bool
+        If True, perform Interleaved ordering in the reverse duration order
     """
     filenames = list(filenames)
     if len(filenames) == 0:
         raise ValueError('No files')
     if len(filenames) == 1:
         return samples_from_file(filenames[0], buffering=buffering, labeled=labeled)
```
Comment (tilmankamp) [inline util/sample_collections.py]: Reversing has also to be done on the interleaved sets - so here `samples_from_file` and (as called from there) `SDB` and `CSV` classes have to do the actual reversing.
```
samples_from_file(filenames[0], buffering=buffering, labeled=labeled, reverse=reverse)
```

File: util/sample_collections.py [hunk 4]
```diff
@@ -355,11 +355,14 @@ def samples_from_files(filenames, buffering=BUFFER_SIZE, labeled=None):
...
     cols = list(map(partial(samples_from_file, buffering=buffering, labeled=labeled), filenames))
```
Comment (tilmankamp) [inline util/sample_collections.py]:
```
cols = list(map(partial(samples_from_file, buffering=buffering, labeled=labeled, reverse=reverse), filenames))
```

File: util/sample_collections.py [hunk 5]
```diff
@@ -355,11 +355,14 @@ def samples_from_files(filenames, buffering=BUFFER_SIZE, labeled=None):
...
-    return Interleaved(*cols, key=lambda s: s.duration)
+    duration_reverse = -1 if reverse else 1
+    return Interleaved(*cols, key=lambda s: duration_reverse * s.duration)
```
Comment (tilmankamp) [inline util/sample_collections.py]: How about
```
return Interleaved(*cols, key=lambda s: -s.duration if reverse else s.duration)
```

File: util/flags.py [hunk 6]
```diff
@@ -90,9 +90,9 @@ def create_flags():
 
     # Sample limits
 
-    f.DEFINE_integer('limit_train', 0, 'maximum number of elements to use from train set - 0 means no limit')
-    f.DEFINE_integer('limit_dev', 0, 'maximum number of elements to use from validation set- 0 means no limit')
-    f.DEFINE_integer('limit_test', 0, 'maximum number of elements to use from test set- 0 means no limit')
+    f.DEFINE_integer('limit_train', None, 'maximum number of elements to use from train set - 0 means no limit')
+    f.DEFINE_integer('limit_dev', None, 'maximum number of elements to use from validation set - 0 means no limit')
+    f.DEFINE_integer('limit_test', None, 'maximum number of elements to use from test set - 0 means no limit')
```
Comment (tilmankamp) [inline util/flags.py]: Why not just going for 0 as the "no limit" case? This could save you some `None` checks.

File: util/feeding.py [hunk 7]
```diff
@@ -109,14 +110,23 @@ def to_sparse_tuple(sequence):
...
                    buffering=1 * MEGABYTE):
+
+    maybe_reverse = False
+    if limit and limit < 0:
+        limit *= -1
+        maybe_reverse = True
+
     def generate_values():
-        samples = samples_from_files(sources, buffering=buffering, labeled=True)
+        samples = samples_from_files(sources, buffering=buffering, labeled=True, reverse=maybe_reverse)
```
Comment (tilmankamp) [inline util/feeding.py]: How about
```samples = samples_from_files(sources, buffering=buffering, labeled=True, reverse=limit < 0)```

File: util/feeding.py [hunk 8]
```diff
@@ -109,14 +110,23 @@ def to_sparse_tuple(sequence):
...
+        if limit and limit > 0:
+            samples = itertools.islice(samples, limit)
```
Comment (tilmankamp) [inline util/feeding.py]: And
```
if abs(limit) > 0:
    samples = itertools.islice(samples, abs(limit))
```
